baseapp/abci.go

```diff
@@ -1264,7 +1258,18 @@ func (app *BaseApp) CreateQueryContext(height int64, prove bool) (sdk.Context, e
 			)
 	}
 
-	cacheMS, err := qms.CacheMultiStoreWithVersion(height)
+	// Acquire an exclusive lock to ensure that the multistore can be versioned.
+	// Internally the multistore updates version ranges while loading which leads to data races.
```

Comment:
Can you explain this a bit more?

Is `cacheStore, err = store.(*iavl.Store).GetImmutable(version)` the problem in `CacheMultiStoreWithVersion`?

Reply:
The problem is with GetVersion since it happens on a mutable tree which tries to load version information from the database mutating first/latest version in its own internal representation. Once the version information is cached then we are working on a store backed by an immutable tree.